### PR TITLE
Modify mode-line lighter customization

### PR DESCRIPTION
**Pull request by sellout**
This adds a new customizable variable, `envrc-lighter-function` that sits between `envrc-lighter` and `envrc-STATUS-lighter`. This allows overriding `envrc-lighter` without relying on internal definitions (`envrc--status`).

The other benefit of this is that any customization is likely to want to set all three `envrc-STATUS-lighter`, with them being mostly the same, with a small conditional part (as you can see from both `envrc--default-lighter` and my lambda below). While this variable is a `risky-local-variable`, I believe that `envrc-STATUS-lighter` are also risky, because they can also contain `:eval`, no?

If `envrc-lighter-function` is not customized, the existing `envrc-STATUS-lighter` variables should continue to work as they have.

In future, I think this could be the _only_ `-lighter` customizable variable.

I like my minor modes to be pretty minimal, so I use this like
```elisp
(setq envrc-lighter-function
      (lambda (status)
        `(:propertize "🗁" ; looks like GH’s font doesn’t have a glyph for “open folder” 
                      face
                      ,(pcase status
                         ('on 'envrc-mode-line-on-face)
                         ('error 'envrc-mode-line-error-face)
                         (_ 'envrc-mode-line-none-face)))))
```

### CODE REVIEW

File: envrc.el [hunk 1]
```diff
@@ -91,21 +91,41 @@ Messages are written into the *envrc-debug* buffer."
...
-(defcustom envrc-lighter '(:eval (envrc--lighter))
+(defcustom envrc-lighter-function #'envrc-standard-lighter
+  "A function that returns a mode-line lighter for the provided status.
+The status is one of the symbols on, error, or none). This variable may also be
+set to nil, which disables the lighter."
+  :type '(choice (const :tag "Disable lighter" nil) function))
+(put 'envrc-lighter-function 'risky-local-variable t)
+
+(defcustom envrc-lighter
+  '(:eval (unless (null envrc-lighter-function)
```
Comment (purcell) [inline envrc.el:102]:
```suggestion
  '(:eval (when envrc-lighter-function
```

**Review by purcell — COMMENTED** [review]
Thanks! I figured people would just write their custom function and set `envrc-lighter` to `(:eval (my-lighter-function)`. Part of me thinks there's no point changing custom vars in order to "unwrap" that function, but in practical terms I don't think anyone is going to set the lighter to a constant string, for example, so I'm +1 on this PR. If we're doing that, I'd also happy to just go ahead and remove `envrc-lighter` and mark it obsolete, plus remove the various `envrc-STATUS-lighter` vars. Would you like to amend the PR to do that too?

**Comment by purcell** [comment]
P.S. Sorry that CI is a bit broken currently, I still haven't figured out what's wrong with the failing tests.

**Comment by sellout** [comment]
tl;dr: This is basically just rambling, no real information here other than “I’ll obsolete the oher vars”.

> Thanks! I figured people would just write their custom function and set `envrc-lighter` to `(:eval (my-lighter-function)`. Part of me thinks there's no point changing custom vars in order to "unwrap" that function, but in practical terms I don't think anyone is going to set the lighter to a constant string, for example, so I'm +1 on this PR.

An alternative I meant to mention in the description is just making `envrc--status` non-internal (`envrc-status`). And then setting `envrc-lighter` directly makes more sense. I also feel like exposing `envrc--status` is an API improvement, perhaps with an `envrc-status-change-hook` so users can make other decisions based on envrc (although I have no immediate use case for this).

> If we're doing that, I'd also happy to just go ahead and remove `envrc-lighter` and mark it obsolete, plus remove the various `envrc-STATUS-lighter` vars.

I think removing `envrc-STATUS-lighter` is the _real_ goal of this PR. Historically, I had configured the lighter like
```elisp
(defun sellout--envrc-lighter (status)
  "Return a lighter for the provided envrc STATUS."
  `(:propertize "🗁"
                face
                ,(pcase status
                   ('error 'envrc-mode-line-error-face)
                   ('on 'envrc-mode-line-on-face)
                   (_ 'envrc-mode-line-none-face))))

(use-package envrc
  :custom
  (envrc-error-lighter (sellout--envrc-lighter 'error))
  (envrc-none-lighter (sellout--envrc-lighter 'none))
  (envrc-on-lighter (sellout--envrc-lighter 'on))
  :hook (after-init . envrc-global-mode))
```
Which I prefer over
```elisp
(use-package envrc
  :custom
  (envrc-lighter '(:eval (sellout--envrc-lighter envrc--status)))
  :hook (after-init . envrc-global-mode))
```
because … I guess just because of `envrc--status` being internal. But, despite me doing that, I don’t see much of a case for wanting to set the `envrc-STATUS-lighter`s individually. I guess I feel like by setting `envrc-lighter` directly, I would be bypassing part of system, making `envrc-STATUS-lighter` useless. So getting rid of `envrc-STATUS-lighter` just reduces my idiosyncratic mental stress 😅

Another possible benefit of `envrc-lighter-function` is that instead of having `envrc-lighter` use `:eval`, `envrc--apply` (or something around there) could `(setq envrc-lighter (envrc-lighter-function envrc--status))`, so that the lighter only gets updated when the environment does, rather than on every mode-line redraw.

> Would you like to amend the PR to do that too?

Yeah, I’m happy to.

**Comment by purcell** [comment]
Yeah, I guess I'm not _super_ averse to making `envrc--status` public, ie. `envrc-status`. Or – potentially safer for the long run – exposing it via a corresponding function.

Regarding `:eval` or not, I'd probably be equally happy with the `envrc-lighter-function` path.

**Comment by purcell** [comment]
I committed something along these lines a while back, and it's evolved further since #130, so I'll close this now. Thanks!
